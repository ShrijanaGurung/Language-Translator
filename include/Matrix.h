/*******************************************************************************
 * Name            : matrix.h
 * Project         : fcal
 * Description     : Header file for matrix
 * Copyright       : 2017 CSCI3081W Staff. All rights reserved.
 * Original Author : Eric Van Wyk
 * Modifications by: John Harwell, Luke Erlandson, Shrijana Gurung
 ******************************************************************************/
#ifndef PROJECT_INCLUDE_MATRIX_H_
#define PROJECT_INCLUDE_MATRIX_H_

#include <stdlib.h>
#include <fstream>
#include <iostream>
#include <string>

class matrix {
 public:
  matrix(int i, int j);
  matrix(const matrix &m);             // Copy Constructor
  matrix &operator=(const matrix &b);  // Copy Assignment

  int n_rows() const;
  int n_cols() const;

  float *access(const int i, const int j) const;
  friend std::ostream &operator<<(std::ostream &os, const matrix &m);

  const matrix operator*(const matrix &b) const;
  const matrix operator+(const matrix &b) const;
  ~matrix();

  static matrix matrix_read(std::string filename);

 private:
  matrix() {}
  int rows;
  int cols;

  /* Your implementation of "data" may vary.  There are ways in
     which data can be an array of arrays and thus simplify the
     access method, at the cost of complicating the process of
     allocating space for data.  The choice is entirely up to
     you. */
  float *data;
};

#endif  // PROJECT_INCLUDE_MATRIX_H_
