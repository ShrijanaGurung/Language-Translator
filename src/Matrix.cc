/*******************************************************************************
 * Name            : Matrix
 * Project         : fcal
 * Module          : matrix
 * Description     : Implementation file for Matrix class
 * Original Author : Luke Erlandson and Shrijana Gurung
 * Copyright       : Luke Erlandson and Shrijana Gurung
 ******************************************************************************/

/*******************************************************************************
 * Includes
 ******************************************************************************/
#include "include/Matrix.h"
#include <cassert>
#include <fstream>
#include <iostream>

/*******************************************************************************
 * Constructors/Methods for our Matrix class
 ******************************************************************************/
/*!
 * Matrix Constructor
 */
matrix::matrix(int i, int j) : rows(i), cols(j) {
  data = reinterpret_cast<float *>(malloc(sizeof(int) * i * j));
}

/*!
 * Matrix Constructor
 */
matrix::matrix(const matrix &m) : rows(m.rows), cols(m.cols) {
  int i, j;
  data = reinterpret_cast<float *>(malloc(sizeof(int) * m.rows * m.cols));
  for (i = 0; i < rows; ++i) {
    for (j = 0; j < cols; ++j) {
      *(access(i, j)) = *(m.access(i, j));
    }
  }
}
/*!
 * Matrix destructor
*/
matrix::~matrix() { delete[] data; }

/*!
 * Copy assignment
 */
matrix &matrix::operator=(const matrix &b) {
  if ( rows == b.n_rows() && cols == b.n_cols() ) {
    data = new float[rows * cols];
    for (int i = 0; i < rows; ++i) {
      for (int j = 0; j < cols; ++j) {
        *(access(i, j)) = *(b.access(i, j));
      }
    }
  } else { std::cout << "Incorrect matrix assignment\n"; }
  return *this;
}

/*
 * Matrix addition
 */
const matrix matrix::operator+(const matrix &b) const {
  int i, j;
  assert(b.n_rows() == rows);
  assert(b.n_cols() == cols);
  matrix c = *this;
  for (i = 0; i < rows; ++i) {
    for (j = 0; j < cols; ++j) {
      *c.access(i, j) += *(b.access(i, j));
    }
  }
  return c;
}
/*!
 * Matrix multiplication
 * */
const matrix matrix::operator*(const matrix &b) const {
  matrix c = matrix(rows, b.cols);
  double sum;
  if (cols == b.rows) {
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < b.cols; j++) {
        sum = 0;
        for (int k = 0; k < b.rows; k++) {
          sum += *this->access(i, k) * *b.access(j, k);
        }
        *c.access(i, j) = sum;
      }
    }
  } else { std::cout << "Incorrect matrix multiplication\n"; }
  return c;
}

/*!
 * Matrix n_rows
 */
int matrix::n_rows() const { return rows; }

/*!
 * Matrix n_cols
 */
int matrix::n_cols() const { return cols; }

/*!
 * Matrix access
 */
float *matrix::access(const int i, const int j) const {
  return data + (i * cols) + j;
}

/*!
 * Matrix matrix_read
 */
matrix matrix::matrix_read(std::string filename) {
  std::ifstream in(filename.c_str());
  if (!in) {
    throw "File " + filename + " not found.";
  }

  int r;
  in >> r;
  int c;
  in >> c;

  matrix temp(r, c);
  for (int i = 0; i < r; i++) {
    for (int j = 0; j < c; j++) {
      in >> *(temp.access(i, j));
    }
  }
  return temp;
}

/*!
 * Matrix ostream
*/
std::ostream &operator<<(std::ostream &os, const matrix &m) {
  os << m.rows << " " << m.cols << std::endl;
  for (int i = 0; i < m.rows; ++i) {
    for (int j = 0; j < m.cols; ++j) {
      os << *(m.access(i, j)) << "  ";
    }
    os << std::endl;
  }
  return os;
}

