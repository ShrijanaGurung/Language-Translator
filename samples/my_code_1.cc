#include <iostream>
 #include "include/Matrix.h"
 #include <math.h>
 using namespace std;
 int main() {
matrix m ( matrix::matrix_read("./samples/my_code.dat"));
 matrix n ( matrix::matrix_read("./samples/my_code2.dat"));
 matrix a (3,5); 
 for( int i=0;i < 3;i++) { 
 for(int j = 0;j< 5;j++) { *(  a.access(i,j)) = 0 ;
}} matrix b (3,3); 
 for( int i=0;i < 3;i++) { 
 for(int j = 0;j< 3;j++) { *(  b.access(i,j)) = 0 ;
}} a = m*b ;
 cout << a ;
 int rows;
 rows = m.n_rows() ;
 int cols;
 cols = m.n_cols() ;
 int i;
 int j;
 int count;
 count = 0 ;
 while(count<2){ for ( i = 0 ; i <= rows-1 ; i++  ) { for ( j = 0 ; j <= cols-1 ; j++  ) { *(m .access( i , j )) = *(m.access(i,j))*2 ;
  }  } count = count+1 ;
  } cout << "New m:\n" ;
 cout << m ;
 b = m*n ;
 cout << "b:\n" ;
 cout << b ;
 }
