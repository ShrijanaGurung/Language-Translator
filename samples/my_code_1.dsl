main () { 
  matrix m = matrix_read ( "./samples/my_code.dat" ) ;
  matrix n = matrix_read ( "./samples/my_code2.dat" ) ;
  matrix a [ 3: 5] i:j = 0;
  matrix b [ 3: 3] i:j = 0;
  a = m * b;
  print ( a );
  int rows;
  rows = n_rows(m);
  
  int cols;
  cols = n_cols(m);
  int i;
  int j;
  int count;
  count = 0;
  while (count < 2){
    repeat (i = 0 to rows - 1){
    repeat (j = 0 to cols - 1){
      m [i : j ] = m [i : j ] * 2;
  }
  }
  count = count + 1;
  }
     
  print ( "New m:\n" ) ;
  print ( m );
  b = m * n;
  print ( "b:\n" ) ;
  print ( b );
}



