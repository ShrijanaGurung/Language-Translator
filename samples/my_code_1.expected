Incorrect matrix multiplication
Incorrect matrix assignment
3 5
0  0  0  0  0  
0  0  0  0  0  
0  0  0  0  0  
New m:
3 4
4  8  12  16  
4  8  12  16  
4  8  12  16  
b:
3 3
72  72  72  
72  72  72  
72  72  72  
